### PR TITLE
[Bug] Solve saved query bug

### PR DESCRIPTION
**Pull request by ananzh**
### Description
In Discover, when loading a saved query with double quotes, double quotes are gone and term query becomes match query. 
Screen Shot 2022-03-07 at 5 40 41 PM

To do a term query, we need to keep the quotes.  When parsing a string with quotes, JSON.parse function will trim the quotes and return its value. For example, `"Men's Shoes"` is parsed to `Men's Shoes`. Then dashboards will break the term using its analyzer and search each word which causes the issue. Please check the [comments](https://github.com/opensearch-project/OpenSearch-Dashboards/issues/1254#issuecomment-1061191659) in the resolved issue for more details. 

To solve this, we added an `isObject` checker which filters a regular string from a json string. After we get the regular string, the query will use its original value and quotes are able to be kept. After change:

Screen Shot 2022-03-07 at 9 13 49 PM

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
    - [x] `yarn test:jest`
    - [x] `yarn test:jest_integration`
    - [x] `yarn test:ftr`
- [ ] New functionality has been documented.
- [x] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: src/plugins/data/public/query/saved_query/saved_query_service.ts [hunk 1]
```diff
@@ -143,20 +144,22 @@ export const createSavedQueryService = (
     id: string;
     attributes: SerializedSavedQueryAttributes;
   }) => {
-    let queryString;
+    let parsedQuery;
```
Comment (kavilla) [inline src/plugins/data/public/query/saved_query/saved_query_service.ts:147]: How come the rename?

Comment (ananzh) [reply src/plugins/data/public/query/saved_query/saved_query_service.ts:147]: the parsed one could be a string or object. I am thinking queryString is a bit ambiguous.

Comment (ananzh) [reply src/plugins/data/public/query/saved_query/saved_query_service.ts:147]: I mean `queryString: string | object`
since it is not necessary a string, should we rename it to avoid confusion?

Comment (ananzh) [reply src/plugins/data/public/query/saved_query/saved_query_service.ts:147]: more tests and explain is here: https://github.com/opensearch-project/OpenSearch-Dashboards/issues/1254#issuecomment-1061191659

File: src/plugins/data/public/query/saved_query/saved_query_service.ts [hunk 2]
```diff
@@ -143,20 +144,22 @@ export const createSavedQueryService = (
...
-      queryString = savedQuery.attributes.query.query;
+      parsedQuery = savedQuery.attributes.query.query;
     }
+
     const savedQueryItems: SavedQueryAttributes = {
       title: savedQuery.attributes.title || '',
       description: savedQuery.attributes.description || '',
       query: {
-        query: queryString,
+        query: isObject(parsedQuery) ? parsedQuery : savedQuery.attributes.query.query,
```
Comment (kavilla) [inline src/plugins/data/public/query/saved_query/saved_query_service.ts:158]: There's some redundancy here that I think can be cleaned up.  

`savedQuery.attributes.query.query` is used three times. Can we define `queryString` as

`const queryString = savedQuery.attributes.query.query` then we can make use of the parsedQuery 
as we try to
```
try {
  parsedQuery = JSON.parse(queryString)
  parsedQuery = isObject(parsedQuery) ? parsedQuery : queryString
 catch (error) {
   parsedQuery = queryString;
 }
```

then `query` can just be updated to take parsedQuery

**Comment by kavilla** [comment]
NIT: The commit message is a little ambiguous.

**Review by kavilla — COMMENTED** [review]
Thanks for handling this! I think it can make to 1.3.0 just some clean ups could be made.

File: src/plugins/data/public/query/saved_query/saved_query_service.ts (hunk 2 above)
Comment (ananzh) [reply src/plugins/data/public/query/saved_query/saved_query_service.ts:158]: I see. will fix to avoid redundancy.

File: src/plugins/data/public/query/saved_query/saved_query_service.test.ts [hunk 3]
```diff
@@ -287,6 +292,175 @@ describe('saved query service', () => {
...
+            query: 'undefined',
+          },
+        },
+      });
+
+      const response = await getSavedQuery('foo');
+      expect(response.attributes.query.query).toEqual('undefined');
+    });
+
+    it('should handle undefined string with double quote', async () => {
```
Comment (ashwin-pc) [inline src/plugins/data/public/query/saved_query/saved_query_service.test.ts:414]: This test case seems wrong, how is it different from the previous test case?

File: src/plugins/data/public/query/saved_query/saved_query_service.test.ts [hunk 4]
```diff
@@ -287,6 +292,175 @@ describe('saved query service', () => {
...
+
+    it('should handle null string with double quote', async () => {
+      mockSavedObjectsClient.get.mockReturnValue({
+        id: 'foo',
+        attributes: {
+          title: 'foo',
+          description: 'bar',
+          query: {
+            language: 'kuery',
+            query: 'null',
```
Comment (ashwin-pc) [inline src/plugins/data/public/query/saved_query/saved_query_service.test.ts:371]: This test seems to be the same as the previous test. Also are these null tests really necessary? seems redundant to have 4 separate tests just to test 2 variations of null and undefined, especially when the original issue had no issue with null or undefined queries but rather with how quotes were escaped.

File: src/plugins/data/public/query/saved_query/saved_query_service.ts (hunk 2 above)
Comment (ashwin-pc) [reply src/plugins/data/public/query/saved_query/saved_query_service.ts:158]: The original issue mentioned two cases which would cause issues
- `'\"Men's Shoes\"`
- `'\"Men's Shoes\" OR \"Men's Clothing\"`

both the existing code and new code both handle the second case identically. This is also the usecase in which the parse error is thrown. Is that intentional?

**Review by ashwin-pc — CHANGES_REQUESTED** [review]
Nice catch! just a few questions, but looks good otherwise

File: src/plugins/data/public/query/saved_query/saved_query_service.test.ts (hunk 3 above)
Comment (ananzh) [reply src/plugins/data/public/query/saved_query/saved_query_service.test.ts:414]: Will remove all unit tests on undefined.

File: src/plugins/data/public/query/saved_query/saved_query_service.test.ts (hunk 4 above)
Comment (ananzh) [reply src/plugins/data/public/query/saved_query/saved_query_service.test.ts:371]: For unit tests on both null and undefined, yes I think it is not necessary to test both edges cases. I will remove tests on undefined.

Yes, the original issue is not about null/undefined. But I think previous getSavedQuery unit test does not cover all query cases 1) a json string which will be parsed to object 2) a regular string with single/double quote which will be parsed to string 3) some edge cases: null, undefined

Since I am changing getSavedQuery to cover all the cases, would like to add tests even the original issue is not considering json string or edge cases.

**Comment by ananzh** [comment]
@ashwin-pc  for the last question, not sure why can't do quote reply. So I will just add a comment to discuss your question. From the code and tests, I think the logic is:

The `savedQuery.attributes.query.query` can only be a json string or a regular string. A json string should be parsed to object and a regular string should be parsed to its value, which is the context of the string. We have no issue with a json string. So we just focus on correctly parsing regular string.

* no single/double quote: no issue. For example, `Men Clothing` will be parsed to `Men Clothing`. OSD do the query using `Men` or `Clothing` based on its standard analyzer.
* with single/double quote: For example `"\"Men's Shoes\""`. For this case, we want to do term query which searches the whole term and does not analyze the search term. Therefore, we need to keep the quote. For `"\"Men's Clothing\" or \"Men's Shoes\""`, yes it raises syntax error and I think it is intentional so `queryString` can be set back. 

```
catch (error) {
      queryString = savedQuery.attributes.query.query;
}
```
However, `"\"Men's Shoes\""` can not use the same logic. The only common thing is that both will not be parsed to object. 

To sum up and answer your question directly, I think it is intentional to design like that.

**Comment by ananzh** [comment]
> NIT: The commit message is a little ambiguous.

will modify the commit message

**Review by kavilla — APPROVED** [review]
Awesome thank you!

**Review by ashwin-pc — APPROVED** [review]
Like the new tests!
